pallets/otc-settlements/src/lib.rs

```diff
@@ -0,0 +1,544 @@
+	fn sort_otcs(block_number: BlockNumberFor<T>) {
+		log::debug!(
+			target: "offchain_worker::sort_otcs",
+			"sort_otcs()");
+
+		// acquire offchain worker lock.
+		let lock_expiration = Duration::from_millis(LOCK_TIMEOUT_EXPIRATION);
+		let mut lock = StorageLock::<'_, Time>::with_deadline(SORTED_ORDERS_LOCK, lock_expiration);
+
+		if Self::try_update_last_block_storage(block_number) {
+			if let Ok(_guard) = lock.try_lock() {
+				let sorted_otcs = StorageValueRef::persistent(OFFCHAIN_WORKER_DATA);
+
+				let mut list = vec![];
+				for (otc_id, otc) in <pallet_otc::Orders<T>>::iter() {
+					let otc_price = FixedU128::checked_from_rational(otc.amount_out, otc.amount_in);
+
+					let route = T::Router::get_route(AssetPair {
+						asset_in: otc.asset_out,
+						asset_out: otc.asset_in,
+					});
+					let router_price_before = T::Router::spot_price_with_fee(&route.clone());
+
+					if let (Some(otc_price), Some(router_price)) = (otc_price, router_price_before) {
+						// otc's with no arb opportunity are at the end of the list and are not sorted
+						let price_diff = otc_price.saturating_sub(router_price);
+						list.push((otc_id, otc_price, router_price, price_diff));
+					}
+				}
+
+				// sort the list by the price diff
+				list.sort_by(|a, b| b.3.partial_cmp(&a.3).unwrap());
+
+				sorted_otcs.set(&list);
+			};
+		}
+	}
+
+	/// Iterate over sorted list of OTCs and try to find arbitrage opportunities.
+	fn settle_otcs() {
+		log::debug!(
+			target: "offchain_worker::settle_otcs",
+			"settle OTCs");
+		// iterate over sorted OTCs
+		let sorted_otcs_storage = StorageValueRef::persistent(OFFCHAIN_WORKER_DATA);
+		let sorted_otcs = sorted_otcs_storage
+			.get::<Vec<SortedOtcsStorageType>>()
+			.unwrap_or_default()
+			.unwrap_or_default();
+
+		for (otc_id, otc_price, router_price_before, _price_diff) in sorted_otcs.iter() {
+			log::debug!(
+			target: "offchain_worker::settle_otcs",
+				"test OTC id {:?} ", otc_id);
+
+			if router_price_before > otc_price {
+				log::debug!(
+			target: "offchain_worker::settle_otcs",
+					"no arb, skipping OTC: {:?}", otc_id);
+				continue;
+			}
+
+			let otc = <pallet_otc::Orders<T>>::get(otc_id).unwrap();
+			let route = T::Router::get_route(AssetPair {
+				asset_in: otc.asset_out,
+				asset_out: otc.asset_in,
+			});
+			let maybe_amount = Self::try_find_trade_amount(*otc_id, route.clone());
+			if let Some(sell_amt) = maybe_amount {
+				log::debug!(
+				target: "offchain_worker::settle_otcs",
+						"Sending TX for OTC id: {:?} amount: {:?}",
+						otc_id,
+						sell_amt
+					);
+				let call = Call::settle_otc_order {
+					otc_id: *otc_id,
+					amount: sell_amt,
+					route,
+				};
+				let _ = SubmitTransaction::<T, Call<T>>::submit_unsigned_transaction(call.into());
+			}
+		}
+	}
+
+	/// Try to find the correct amount to close the arbitrage opportunity.
+	fn try_find_trade_amount(otc_id: OrderId, route: Vec<Trade<AssetIdOf<T>>>) -> Option<Balance> {
```

Comment:
you can use slice instead of vec, than there is not need to copy the route where you use this function.

you can also pass the otc by reference, since you already have it , so no need to query storage again inside this function